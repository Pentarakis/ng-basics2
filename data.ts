export const BOOKS = [
  {id: 3, name: 'A Storm of Swords', isbn: '978-0553106633'},
  {id: 4, name: 'The Hedge Knight', isbn: '978-0976401100'},
  {id: 5, name: 'A Feast for Crows', isbn: '978-0553801507'}
];

export const CHARACTERS = [
  { id: 1, name: 'Daenerys Targaryen', culture: 'Valyrian'},
  { id: 2, name: 'Jon Snow', culture: 'Northmen'}
];
