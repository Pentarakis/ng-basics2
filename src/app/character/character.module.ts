import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { MatButtonModule, MatCardModule, MatFormFieldModule, MatInputModule, MatTableModule } from '@angular/material';
import { RouterModule } from '@angular/router';
import { CharacterListComponent } from './character-list/character-list.component';
import { CharacterRoutingModule } from './character-routing.module';
import { CharacterComponent } from './character/character.component';

@NgModule({
  declarations: [CharacterComponent, CharacterListComponent],
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MatTableModule,
    MatCardModule,
    MatFormFieldModule,
    MatInputModule,
    MatButtonModule,
    RouterModule,
    CharacterRoutingModule
  ]
})
export class CharacterModule { }
