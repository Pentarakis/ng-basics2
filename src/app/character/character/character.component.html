<mat-card>

  <mat-card-header>Edit Character</mat-card-header>

  <mat-card-content>
    <form>

      <mat-form-field>
        <input matInput name="name" placeholder="Name" [(ngModel)]="character.name" >
      </mat-form-field>

      <br>

      <mat-form-field>
        <input matInput name="culture" placeholder="Culture" [(ngModel)]="character.culture" >
      </mat-form-field>

      <br>

      <div>
        <button mat-raised-button color="primary" (click)="save()">
          {{ character.id ? 'New' : 'Add' }}
        </button>
      </div>

    </form>
  </mat-card-content>

</mat-card>

