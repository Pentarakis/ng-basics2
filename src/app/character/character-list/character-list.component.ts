import { Component, OnInit } from '@angular/core';
import { Book } from '../../book/model/book';
import { Character } from '../model/character';

@Component({
  selector: 'ngb-character-list',
  templateUrl: './character-list.component.html',
  styleUrls: ['./character-list.component.scss']
})
export class CharacterListComponent implements OnInit {

  characters: Character[] = [
    { id: 1, name: 'Daenerys Targaryen', culture: 'Valyrian'},
    { id: 2, name: 'Jon Snow', culture: 'Northmen'}
  ];

  displayedColumns: string[] = ['id', 'name', 'culture'];

  constructor() { }

  ngOnInit() {
  }

  showDetails(character: Character): void {
  }

}
