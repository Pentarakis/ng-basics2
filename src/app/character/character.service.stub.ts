import { Observable, Subject } from 'rxjs';
import { Character } from './model/character';

export class CharacterServiceStub {

  constructor() { }

  responseSub$: Subject<Character> = new Subject<Character>();
  readAllSub$: Subject<Character[]> = new Subject<Character[]>();

  readAll(): Observable<Character[]> {
    return this.readAllSub$;
  }

  read(id: number): Observable<Character> {
    return this.responseSub$;
  }

  create(character: Character): Observable<Character> {
    return this.responseSub$;
  }

  update(character: Character): Observable<Character> {
    return this.responseSub$;
  }
}
