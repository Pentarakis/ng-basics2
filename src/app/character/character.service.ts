import { Injectable } from '@angular/core';
import { Character } from './model/character';

@Injectable({
  providedIn: 'root'
})
export class CharacterService {

  private readonly baseUrl = 'http://localhost:3000/characters';

  private readonly characters: Character[] = [
    { id: 1, name: 'Daenerys Targaryen', culture: 'Valyrian'},
    { id: 2, name: 'Jon Snow', culture: 'Northmen'}
  ];

  constructor() { }

  readAll(): Character[] {
    return this.characters;
  }

  read(id: number): Promise<Response> {
    return fetch(`${this.baseUrl}/${id}`);
  }
}
