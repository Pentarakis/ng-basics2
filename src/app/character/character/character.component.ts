import { Component, OnInit } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { first } from 'rxjs/internal/operators/first';
import { CharacterService } from '../character.service';
import { Character } from '../model/character';

@Component({
  selector: 'ngb-character',
  templateUrl: './character.component.html',
  styleUrls: ['./character.component.scss']
})
export class CharacterComponent implements OnInit {

  isCreateMode = true;
  character: Character = new Character();

  constructor(private route: ActivatedRoute, private characterService: CharacterService) {

    this.route.params.pipe(first()).toPromise()
      .then(params => {
        const id = params.id;
        if (id !== 'create') {
          this.characterService.read(Number(id))
            .then(response => response.json())
            .catch(() => {
              alert('Communication Error');
              setTimeout(() => {
                this.characterService.read(Number(id))
                  .then(response => response.json())
                  .then(character => this.character = character);
              }, 2000);
            })
            .then(character => this.character = character);
        }

      });
  }

  ngOnInit() {
  }

  save(): void {
  }
}
