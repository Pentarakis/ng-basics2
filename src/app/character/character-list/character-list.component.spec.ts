import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { MatButtonModule, MatTableModule } from '@angular/material';
import { RouterTestingModule } from '@angular/router/testing';
import { CHARACTERS } from '../../../../data';
import { CharacterService } from '../character.service';
import { CharacterServiceStub } from '../character.service.stub';

import { CharacterListComponent } from './character-list.component';

describe('CharacterListComponent', () => {
  let component: CharacterListComponent;
  let fixture: ComponentFixture<CharacterListComponent>;
  let characterService;

  beforeEach(async(() => {
    characterService = new CharacterServiceStub();
    jest.spyOn(characterService, 'read');

    TestBed.configureTestingModule({
      declarations: [ CharacterListComponent ],
      imports: [MatTableModule, MatButtonModule, RouterTestingModule],
      providers: [
        { provide: CharacterService, useValue: characterService }
      ]
    });
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CharacterListComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    characterService.readAllSub$.next(CHARACTERS);
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render MatTable', () => {
    expect(fixture).toMatchSnapshot(); // This is the snapshot validation
  });
});
