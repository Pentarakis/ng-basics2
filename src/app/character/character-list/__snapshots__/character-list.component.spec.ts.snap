// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterListComponent should render MatTable 1`] = `
<ngb-character-list
  characterService={[Function CharacterServiceStub]}
  characters={[Function Subject]}
  displayedColumns={[Function Array]}
>
  <button
    class="mat-raised-button mat-button-base mat-primary"
    color="primary"
    mat-raised-button=""
    routerlink="create"
    tabindex="0"
  >
    <span
      class="mat-button-wrapper"
    >
       Create
  
    </span>
    <div
      class="mat-button-ripple mat-ripple"
      matripple=""
    />
    <div
      class="mat-button-focus-overlay"
    />
  </button><br /><br /><table
    class="mat-elevation-z8 mat-table"
    mat-table=""
    role="grid"
  >
    <thead
      role="rowgroup"
    >
      
      <tr
        class="mat-header-row"
        mat-header-row=""
        role="row"
      >
        
        <th
          class="mat-header-cell cdk-column-id mat-column-id"
          mat-header-cell=""
          role="columnheader"
        >
           ID 
        </th>
        <th
          class="mat-header-cell cdk-column-name mat-column-name"
          mat-header-cell=""
          role="columnheader"
        >
           Name 
        </th>
        <th
          class="mat-header-cell cdk-column-culture mat-column-culture"
          mat-header-cell=""
          role="columnheader"
        >
           Culture 
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      
      <tr
        class="mat-row"
        mat-row=""
        role="row"
        tabindex="0"
      >
        
        <td
          class="mat-cell cdk-column-id mat-column-id"
          mat-cell=""
          role="gridcell"
        >
           
        </td>
        <td
          class="mat-cell cdk-column-name mat-column-name"
          mat-cell=""
          role="gridcell"
        >
           
        </td>
        <td
          class="mat-cell cdk-column-culture mat-column-culture"
          mat-cell=""
          role="gridcell"
        >
           
        </td>
      </tr>
      <tr
        class="mat-row"
        mat-row=""
        role="row"
        tabindex="0"
      >
        
        <td
          class="mat-cell cdk-column-id mat-column-id"
          mat-cell=""
          role="gridcell"
        >
           
        </td>
        <td
          class="mat-cell cdk-column-name mat-column-name"
          mat-cell=""
          role="gridcell"
        >
           
        </td>
        <td
          class="mat-cell cdk-column-culture mat-column-culture"
          mat-cell=""
          role="gridcell"
        >
           
        </td>
      </tr>
    </tbody>
    <tfoot
      class="mat-table-sticky"
      role="rowgroup"
      style="display: none; bottom: 0px; position: sticky; z-index: 10;"
    >
      
    </tfoot>
  </table>
</ngb-character-list>
`;
